File: examples/shared/go/interactive/step.go
package interactive

import (
	"fmt"
	"os"
)

// StepAction represents the result of a step pause.
type StepAction int

const (
	// ActionContinue means user pressed Enter, proceed to next step
	ActionContinue StepAction = iota
	// ActionQuit means user typed 'q', exit program gracefully
	ActionQuit
	// ActionSkip means user typed 's', disable step mode and continue
	ActionSkip
)

// StepPause pauses for step-through mode with explanation.
//
// Displays a title and explanation bullet points, then waits for user input.
// Returns the action the user chose.
//
// Parameters:
//   - title: Brief description of the step (e.g., "Creating Claude provider...")
//   - explanation: Bullet points explaining what will happen
//
// Returns:
//   - ActionContinue: User pressed Enter
//   - ActionQuit: User typed 'q'
//   - ActionSkip: User typed 's'
func (c *Config) StepPause(title string, explanation []string) StepAction {
	if !c.IsStep() {
		return ActionContinue
	}

	// Print step header
	fmt.Fprintf(os.Stderr, "\n[nxusKit STEP] %s\n", title)

	// Print explanation bullets
	for _, line := range explanation {
		fmt.Fprintf(os.Stderr, "  - %s\n", line)
	}

	// Print prompt
	fmt.Fprintln(os.Stderr, "[Press Enter to continue, 'q' to quit, 's' to skip steps]")

	// Read user input
	input := ReadLine()

	switch input {
	case "q", "quit":
		return ActionQuit
	case "s", "skip":
		c.SkipSteps()
		fmt.Fprintln(os.Stderr, "[nxusKit] Step mode disabled, running to completion...")
		return ActionSkip
	default:
		return ActionContinue
	}
}

File: examples/shared/rust/src/step.rs
//! Step-through mode for learning and debugging.

use crate::InteractiveConfig;
use crate::tty::read_line;

/// Result of a step pause, indicating user's choice.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum StepAction {
    /// User pressed Enter, proceed to next step
    Continue,
    /// User typed 'q', exit program gracefully
    Quit,
    /// User typed 's', disable step mode and continue
    Skip,
}

impl InteractiveConfig {
    /// Pause for step-through mode with explanation.
    ///
    /// Displays a title and explanation bullet points, then waits for user input.
    /// Returns the action the user chose.
    ///
    /// # Arguments
    /// * `title` - Brief description of the step (e.g., "Creating Claude provider...")
    /// * `explanation` - Bullet points explaining what will happen
    ///
    /// # Returns
    /// * `StepAction::Continue` - User pressed Enter
    /// * `StepAction::Quit` - User typed 'q'
    /// * `StepAction::Skip` - User typed 's'
    pub fn step_pause(&mut self, title: &str, explanation: &[&str]) -> StepAction {
        if !self.is_step() {
            return StepAction::Continue;
        }

        // Print step header
        eprintln!("\n[nxusKit STEP] {}", title);

        // Print explanation bullets
        for line in explanation {
            eprintln!("  - {}", line);
        }

        // Print prompt
        eprintln!("[Press Enter to continue, 'q' to quit, 's' to skip steps]");

        // Read user input
        match read_line() {
            Some(input) => {
                if input == "q" || input == "quit" {
                    StepAction::Quit
                } else if input == "s" || input == "skip" {
                    self.skip_steps();
                    eprintln!("[nxusKit] Step mode disabled, running to completion...");
                    StepAction::Skip
                } else {
                    StepAction::Continue
                }
            }
            None => {
                // If we can't read input, continue automatically
                StepAction::Continue
            }
        }
    }
}
